immutable.*Map computes hashcode without intermediate Tuple2 creation

mkeskells opened the pull request:
Scala 2.13.x improves the efficiency of `immutable.HashMap#hashCode` in two ways: temporary `Tuple2` instances are [not created during iteration](https://github.com/scala/scala/blob/456d3ef425019870328a2fd4639afab6e9496ebd/src/library/scala/collection/immutable/HashMap.scala#L2089-L2110) when computing the hash code of the map, and the hash codes of keys are stored during map creation.

This PR backports the first improvement to 2.12.x _and_ extends it to the low-arity specialized `immutable.{MapN, TreeMap, ListMap}. 

This doesnt affect the hashcode result, it just uses less tuples.

mkeskells commented:
cc: @retronym

retronym commented:
```
$ scala-ref 2.12.x -nc -e 'println(collection.immutable.HashMap("" -> "").hashCode)'
43880933

$ scala-pr 8399 -nc -e 'println(collection.immutable.HashMap("" -> "").hashCode)'
43880933

```

👍

mkeskells commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -51,6 +51,24 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
 
   override def foreach[U](f: ((A, B)) => U): Unit = ()
+  private[immutable] def foreachKV(f: (A, B) => Unit): Unit = ()
+  override def hashCode(): Int = {
```
maybe this method should be on Map as a helper

mkeskells commented:
All these method that fail Mima are private[scala] or more private
What am I missing?

```
[error]  * method foreachKV(scala.Function2)Unit in class scala.collection.immutable.HashMap does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.HashMap.foreachKV")
[error]  * method foreachKV(scala.collection.immutable.RedBlackTree#Tree,scala.Function2)Unit in object scala.collection.immutable.RedBlackTree does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.RedBlackTree.foreachKV")
[error]  * method foreachKV(scala.Function2)Unit in class scala.collection.immutable.HashMap#HashTrieMap does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.HashMap#HashTrieMap.foreachKV")
[error]  * method foreachKV(scala.Function2)Unit in class scala.collection.immutable.ListMap does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.ListMap.foreachKV")
[error]  * method foreachKV(scala.Function2)Unit in class scala.collection.immutable.HashMap#HashMap1 does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.HashMap#HashMap1.foreachKV")
[error]  * method foreachKV(scala.Function2)Unit in class scala.collection.immutable.HashMap#HashMapCollision1 does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.collection.immutable.HashMap#HashMapCollision1.foreachKV")
[error]  * method product2Hash(java.lang.Object,java.lang.Object)Int in object scala.util.hashing.MurmurHash3 does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.util.hashing.MurmurHash3.product2Hash")
[error]  * method product2Hash(java.lang.Object,java.lang.Object,Int)Int in class scala.util.hashing.MurmurHash3 does not have a correspondent in other version
[error]    filter with: ProblemFilters.exclude[DirectMissingMethodProblem]("scala.util.hashing.MurmurHash3.product2Hash")
```

mkeskells commented:
need to add Map1 - 4

mkeskells commented:
should add non-allocating unit test

retronym commented on src/library/scala/collection/immutable/HashMap.scala:
```diff
@@ -166,6 +173,7 @@ object HashMap extends ImmutableMapFactory[HashMap] with BitOperations.Int {
   private object EmptyHashMap extends HashMap[Any, Nothing] { 
     override def head: (Any, Nothing) = throw new NoSuchElementException("Empty Map")
     override def tail: HashMap[Any, Nothing] = throw new NoSuchElementException("Empty Map")
+    override val hashCode: Int = super.hashCode()
```
This val changes the automatically generated serial version UID.

```
➜  scala-2.12.x git:(topic/tsan-cleanup) ✗ scala-launch 2.12.9 -nc -e 'println(java.io.ObjectStreamClass.lookup(scala.collection.immutable.HashMap.empty[Any, Any].getClass))'
scala.collection.immutable.HashMap$EmptyHashMap$: static final long serialVersionUID = 2121866088229452155L;
➜  scala-2.12.x git:(topic/tsan-cleanup) ✗ scala-pr 8399 -nc -e 'println(java.io.ObjectStreamClass.lookup(scala.collection.immutable.HashMap.empty[Any, Any].getClass))'
scala.collection.immutable.HashMap$EmptyHashMap$: static final long serialVersionUID = -2471795291674749849L;
```

In 2.13.x we consistently use the serialization proxy pattern to avoid this instability. 

For 2.12.x, you could try:
  - marking this val as `@transient`, or
  - revert this line, or
  - replace it with a def returning a hard-coded constant that is proven to be the right value by a unit test

mkeskells replied:
took the third option - a def and a test

mkeskells committed:
Dont create tuples when generating hashcodes for immutable maps
add allocation tests for immutable maps
add pre-calculated values for empty maps, with test

for immutable maps the max allocation is 32 bytes

retronym commented:
I have some review suggestions in https://github.com/retronym/scala/pull/71

  - Add `foreachKeyValue` rather than `includeHash`.
  - Add necessary MiMa entries.

retronym commented:
Submitted an updated version as #8411
